make `wasm32` compatible

DougAnderson444 opened the pull request:
While trying to use `multikey` in WebAssembly code, I realized `sodiumoxide` doesn't compile to that target, and should be upgraded to the RustCrypto crates as it uses the older 8 bytes nonce instead of the latest 12 byte nonce.

I took some liberty to make some small clippy clean ups too.

I changed the serde tests, but didn't look too deeply into the details as they seem a bit fragile with the hard-coded checks in there. Recommend those serde tests be made into something a bit more robust eventually, like roundtrip or somehow codify the encoding to automate the checking. Anyway, the tests pass but should be looked at further I think.

In the meanime, this fixes issue #1 and compiles to wasm32 =)

DougAnderson444 committed:
replace `sodiumoxide` with `chacha20` and `poly1305` crates

fixes issue #1

DougAnderson444 committed:
make ssh-key optional

so that multikey can compile to wasm32 target

DougAnderson444 committed:
start multisig patch

DougAnderson444 committed:
place `ssh` behind feature flag

DougAnderson444 commented:
I also ran into the issue of using `ssh-key` and `ssh-encoding`.

This cannot compile to `wasm32*` either, so I placed it behind a feature flag.

I also switch `new_from_random_bytes` to using `dalek` directly to remove it's dependency on `ssh-key` which is not needed.

DougAnderson444 committed:
Merge branch 'switch-chacha20' into feature-ssh

DougAnderson444 committed:
use ed25519_dalek instead of ssh-key

DougAnderson444 committed:
make deps direct to douganderson444 repo branch

DougAnderson444 commented:
TODO: need to exclude default features on `ssh-key` & `encoding`, as I suspect that's what blocking the wasm build

DougAnderson444 committed:
update gh actions CI

DougAnderson444 committed:
`ssh-key` no-default-features for `wasm32-*` builds

DougAnderson444 commented:
When https://github.com/cryptidtech/multisig/pull/1 get merged, I'll update this PR and make it no longer Draft

DougAnderson444 committed:
rm dup runs-on

DougAnderson444 committed:
Update rust.yml

DougAnderson444 committed:
allow manual github workflow dispatch

DougAnderson444 committed:
add getrandom/js dep for CI build

dhuseby commented:
Your multisig PR was just landed. Flip this one out of Draft so we can move forward.

DougAnderson444 committed:
switch back to cryptid repo

DougAnderson444 committed:
Merge remote-tracking branch 'upstream/main' into wasm32-pure-doug

DougAnderson444 committed:
chacha20poly1305 is 32 byte key, 12 bytes nonce

DougAnderson444 committed:
rm println

DougAnderson444 committed:
comment about why wasm feature is present

DougAnderson444 commented:
I updated to test `test_serde_encrypted_secret_key_compact` to reflect the new `12 byte` chacha2-poly1305 nonce and `32 byte` key.

I added various target checks in the CI, including `wasm32`

All tests passing now, should be good to go!

DougAnderson444 committed:
fix comment typo

DougAnderson444 committed:
Merge remote-tracking branch 'upstream/main' into switch-chacha20

DougAnderson444 committed:
Revert "Merge remote-tracking branch 'upstream/main' into switch-chacha20"

This reverts commit f9746b714f7a84952035e362a74f61e7fa7a5e02, reversing
changes made to 9e11324dc2779aa436b9bfbb0e57eb34884a3863.

DougAnderson444 commented:
Github was showing conflicts, which I tried to resolve and now there's no conflicts showing. Weird.

Recommend running the CI again to confirm it's still good to merge.

DougAnderson444 committed:
redo clippy fixes

Somewhere in my merge and revert, the clippy fixes were lost. This commit ensures all the clippy corrections are back. Git!

dhuseby commented on src/mk.rs:
```diff
@@ -93,17 +93,29 @@ impl EncodingInfo for Multikey {
+impl Into<Vec<u8>> for Multikey {
+    fn into(self) -> Vec<u8> {
+impl From<Multikey> for Vec<u8> {
+    fn from(mk: Multikey) -> Vec<u8> {
```
This looks incorrect. It looks like a merge/clippy --fix error.

DougAnderson444 replied:
I thought I caught all these but looks like I missed some. When I did the merge something went wrong and it kept old and new copies. Sigh.

The [`from(value: T)`](https://doc.rust-lang.org/std/convert/trait.From.html#tymethod.from) is the idiomatic way we should be doing it to be consistent with the trait IMO, but it's a bikeshed color so I'm fine either way.

I'll go through these and clean them up

DougAnderson444 replied:
These are marked outdated, did you make more commits? I'll take a look either way

DougAnderson444 replied:
Yeah it looks like these dupes are from an older commit, the latest commit should be at https://github.com/cryptidtech/multikey/pull/2/commits/3b89dc2bc7b9b5fc1f6f53b1c29d2cc6d227d3ec

DougAnderson444 replied:
You may have not pulled down those last couple of commits, which would make the difference

DougAnderson444 replied:
Fixed in https://github.com/cryptidtech/multikey/pull/2/commits/f479bd484b9c377a0a44684ec0b4160d04db1b71

dhuseby replied:
oh shit...sorry...I didn't do a force push though so I wasn't aware that there were commits I needed to pull first. I'm sorry 'bout that.

dhuseby commented on src/mk.rs:
```diff
@@ -93,17 +93,29 @@ impl EncodingInfo for Multikey {
         // add in the key codec
         v.append(&mut val.codec.into());
+        v.append(&mut self.codec.clone().into());
+        v.append(&mut mk.codec.into());
```
Same here.

DougAnderson444 replied:
Fixed in https://github.com/cryptidtech/multikey/pull/2/commits/f479bd484b9c377a0a44684ec0b4160d04db1b71

dhuseby commented on src/mk.rs:
```diff
@@ -93,17 +93,29 @@ impl EncodingInfo for Multikey {
         // add in the codec-specific attributes
         val.attributes.iter().for_each(|(id, attr)| {
+        self.attributes.iter().for_each(|(id, attr)| {
+        mk.attributes.iter().for_each(|(id, attr)| {
```
and this.

DougAnderson444 replied:
I think you may be looking at an old version, Github has these marked as outdated and I don't see these issues on my branch.

I think I may have already resolved these, which is good because it means I'm not losing my mind

DougAnderson444 replied:
Fixed as of https://github.com/cryptidtech/multikey/pull/2/commits/f479bd484b9c377a0a44684ec0b4160d04db1b71

dhuseby commented on src/mk.rs:
```diff
@@ -404,14 +416,20 @@ impl Builder {
             Codec::P521Priv => EcdsaKeypair::random(rng, EcdsaCurve::NistP521)
                 .map_err(|e| ConversionsError::Ssh(e.into()))?
+                .map_err(|e| ConversionsError::SshKey(e))?
+                .map_err(ConversionsError::SshKey)?
```
and this.

DougAnderson444 replied:
Fixed as of https://github.com/cryptidtech/multikey/pull/2/commits/f479bd484b9c377a0a44684ec0b4160d04db1b71

dhuseby commented on src/views/bls12381.rs:
```diff
@@ -491,19 +497,25 @@ impl<'a> ConvView for View<'a> {
                     .encode(&mut buf)
                     .map_err(|e| ConversionsError::Ssh(e.into()))?;
+                    .map_err(|e| ConversionsError::SshEncoding(e))?;
+                    .map_err(ConversionsError::SshEncoding)?;
```
same.

DougAnderson444 replied:
Fixed as of https://github.com/cryptidtech/multikey/pull/2/commits/f479bd484b9c377a0a44684ec0b4160d04db1b71

dhuseby commented on src/views/bls12381.rs:
```diff
@@ -623,6 +657,8 @@ impl<'a> ConvView for View<'a> {
         )
         .map_err(|e| ConversionsError::Ssh(e.into()))?)
+        .map_err(|e| ConversionsError::SshKey(e))?)
+        .map_err(ConversionsError::SshKey)?)
```
same

DougAnderson444 replied:
Fixed as of https://github.com/cryptidtech/multikey/pull/2/commits/f479bd484b9c377a0a44684ec0b4160d04db1b71

dhuseby reviewed (changes requested):
It looks like there was a merge/clippy --fix error in all of these changes. Duplicated code everywhere. BTW, I like `fn from(mk: Multikey)` better than `fn from(val: Multikey)`. This may be my fault because I was pushing clippy fixes this week but it's in your PR so you get to fix it 😂

DougAnderson444 committed:
use `from(mk: Multikey)`

because Dave likes it better

dhuseby commented on Cargo.toml:
```diff
@@ -1,6 +1,6 @@
 [package]
 name = "multikey"
-version = "1.0.7"
+version = "1.0.6"
```
Are we downgrading for a reason?

DougAnderson444 replied:
Definitely not, that's weird. Easy fix.

dhuseby commented on src/cipher.rs:
```diff
@@ -94,7 +94,8 @@ mod tests {
 
-        let nonce = hex::decode("00b61a43d4d1e8d7").unwrap();
+        // ChaCha needs 12 bytes of nonce iaw RFC8439
+        let nonce = hex::decode("00b61a43d4d1e8d700b61a43").unwrap();
```
Good catch

dhuseby commented on src/error.rs:
```diff
@@ -124,6 +118,48 @@ pub enum ConversionsError {
 
+/// SSH Encoding Errors that cannot be handled by thiserror since they may not use the std feature
+/// in the case of wasm32 target.
+#[derive(Clone, Debug)]
```
if you add thiserror::Error you don't have to do the impls. See this file as an example: https://github.com/cryptidtech/provenance-log/blob/main/src/error.rs

DougAnderson444 replied:
Ah yes, good point! Fixed.

DougAnderson444 replied:
Sure, `thiserror` works great when `ssh-key` has `std` feature enabled.

But when disabled, as in this PR, `StdError` and `AsDynError` are not implement on the `ssh_key` or `ssh_encoding` creates, which cause the `wasm32` builds to fail.

Try doing this with those `thiserror` changes:

```
cargo build --target wasm32-unknown-unknown --features wasm
```

Since `thiserror` cannot be used, they are implemented as you see, with all the `e.into()` conversions to get them from `SshError` into a `ConversionError`. `Display` transparently passes the source up, same as `thiserror`. 

The code is correct in order to build to `wasm32`.

dhuseby replied:
If we can't use thiserror then how does ConversionError compile?

dhuseby commented on src/error.rs:
```diff
@@ -124,6 +118,48 @@ pub enum ConversionsError {
+/// in the case of wasm32 target.
+#[derive(Clone, Debug)]
+pub enum SshErrors {
+    Key(ssh_key::Error),
```
also need to use #[error] on these

DougAnderson444 replied:
yep, done as part of previous thiserrorama

DougAnderson444 replied:
`thiserror` does not work without `std` feature for `ssh-key` for `wasm32` builds

dhuseby commented on src/mk.rs:
```diff
@@ -32,20 +32,19 @@ pub const KEY_CODECS: [Codec; 4] = [
+                                Codec::LamportSha3256PrivShare,
+                                Codec::LamportSha3384PrivShare,
+                                Codec::LamportSha3512PrivShare,
+                                */
```
👍

dhuseby commented on src/mk.rs:
```diff
@@ -402,17 +401,17 @@ impl Builder {
     ) -> Result<Self, Error> {
         let key_bytes = match codec {
-            Codec::Ed25519Priv => Ed25519Keypair::random(rng).private.to_bytes().to_vec(),
+            Codec::Ed25519Priv => ed25519_dalek::SigningKey::generate(rng).to_bytes().to_vec(),
```
What was wrong with using Ed25519Keypair?

DougAnderson444 replied:
When I made this change, I was in the process of placing `ssh` behind a feature flag, the change was needed otherwise users didn't have access to Ed25519 since `Ed25519Keypair` comes from `ssh_key::private::Ed25519Keypair`. `ssh-key` sources from `dalek` anyway, so it's the same dep. 

Once I shifted to the `ssh-key defaults = false` instead of an `ssh` feature flag, that reason is no longer valid. 

That said, if this crate ever uses feature flags for `ssh` (maybe it still should? not everyone may want the extra dependency?) then the replacement would be valid again.

I don't see an impact either way at this point, pending further discussion on feature flags.

dhuseby commented on src/mk.rs:
```diff
@@ -402,17 +401,17 @@ impl Builder {
+            Codec::Ed25519Priv => ed25519_dalek::SigningKey::generate(rng).to_bytes().to_vec(),
             Codec::P256Priv => EcdsaKeypair::random(rng, EcdsaCurve::NistP256)
-                .map_err(ConversionsError::SshKey)?
+                .map_err(|e| ConversionsError::Ssh(e.into()))?
```
These are incorrect. If you use thiserror::Error in deriving the Ssh errors, you won't need e.into()

DougAnderson444 replied:
Code is correct. `thiserror` does not work without `std` feature for `ssh-key` for `wasm32` builds

dhuseby commented on src/mk.rs:
```diff
@@ -819,7 +818,7 @@ impl Builder {
     fn with_attribute(mut self, attr: AttrId, data: &Vec<u8>) -> Self {
         let mut attributes = self.attributes.unwrap_or_default();
-        attributes.insert(attr, data.to_owned().into());
+        attributes.insert(attr, data.clone().into());
```
Also, to_owned() is correct here because it's a reference that is passed in.

DougAnderson444 replied:
I don't recall making this change, but it's fixed now

dhuseby commented on src/nonce.rs:
```diff
@@ -126,7 +121,8 @@ pub struct Builder {
     /// build from random source
     pub fn new_from_random_bytes(size: usize, rng: &mut (impl RngCore + CryptoRng)) -> Self {
-        let mut bytes = vec![0; size];
+        let mut bytes = Vec::with_capacity(size);
```
This is incorrect. My clippy says to use the vec! macro. I assume because it allows for faster code due to the macro expansion at compile time.

DougAnderson444 replied:
Yeah, mine too. Fixed!

dhuseby commented on src/views/bls12381.rs:
```diff
@@ -167,19 +167,27 @@ impl<'a> AttrView for View<'a> {
 
     fn is_secret_key(&self) -> bool {
-        matches!(self.mk.codec,
+        match self.mk.codec {
```
this is also incorrect. matches! makes faster code.

DougAnderson444 replied:
We like fast! Fixed!

dhuseby commented on src/views/bls12381.rs:
```diff
@@ -456,7 +464,7 @@ impl<'a> ConvView for View<'a> {
                 key_bytes
                     .encode(&mut buf)
-                    .map_err(ConversionsError::SshEncoding)?;
+                    .map_err(|e| ConversionsError::Ssh(e.into()))?;
```
if you fix the SshError derivation to include thiserror::Error you won't need these.

DougAnderson444 replied:
can't use `thiserror` for `wasm32`, as per previous

dhuseby commented on src/views/chacha20.rs:
```diff
@@ -94,15 +99,15 @@ impl<'a> CipherAttrView for View<'a> {
+        let nonce =
+            Nonce::from_exact_iter(nonce.iter().copied()).ok_or(CipherError::InvalidNonce)?;
+
+        Ok(nonce.to_vec().into())
```
This is great! 👍

DougAnderson444 replied:
Thanks! 🤗

dhuseby reviewed (changes requested):
Fix the SshError derivation to use thiserror and a lot of the changes go away because the error elision is possible at compile time. There's also a few changes you made that are incorrect. I think using standard macros like vec! and matches! produces faster code IIRC.

DougAnderson444 committed:
v1.0.7

DougAnderson444 committed:
use matches!

DougAnderson444 committed:
use vec!, and other clippy and rustfmt chgs

I don't recall the ratioanle for this change, but now it's fixed

DougAnderson444 committed:
add missing rustdoc comments

DougAnderson444 committed:
use `to_owned()`

DougAnderson444 commented:
@dhuseby I believe I have addressed and changed all your points, can you have another look with the latest commits?

DougAnderson444 committed:
multikey from seed + test (Ed25519 only for now)

dhuseby reviewed (approved):
LGTM

dhuseby commented:
🚀
